fem/coefficient.cpp

```diff
@@ -294,6 +294,8 @@ void VectorRestrictedCoefficient::Eval(
 void MatrixFunctionCoefficient::Eval(DenseMatrix &K, ElementTransformation &T,
                                      const IntegrationPoint &ip)
 {
+   MFEM_VERIFY(!symmetric, "MatrixFunctionCoefficient is symmetric");
```

Comment:
We should be able to support the symmetric case as well.  Users might have a reason for not calling EvalSymmetric.

Reply:
@mlstowell `Eval` sets a `DenseMatrix`, whereas `EvalSymmetric` sets a `Vector` of the upper triangular entries. The flag `symmetric` does not necessarily mean that the matrix is symmetric; the real meaning is that the `MatrixCoefficient` is intended to be evaluated via `EvalSymmetric`, e.g. a function has been provided for computing the upper triangular entries. 

The user could call `Eval` for a symmetric `MatrixFunctionCoefficient` created with the constructor that takes a function defining all matrix entries. That constructor sets `MatrixCoefficient::symmetric` to `false`, by default, so that `Eval` can be called. It does not matter whether the matrix is actually symmetric mathematically, just whether it is specified by a function that computes all entries or just the upper triangular entries.

This summarizes my intended use of these functions, but your comment is still valid. We could decide to support the symmetric case in `MatrixFunctionCoefficient`, where `symmetric` being true means that `SymmFunction` should be called, not `Function` or `TDFunction`. In that case, `Eval` should call `SymmFunction` to get a `Vector` of upper triangular values, and then insert those into the full `DenseMatrix` `K`. This would be a more complete implementation, allowing the user to call `Eval`, but it would be less efficient than calling `EvalSymmetric`. It is a design question, of whether we want to allow the user to call the less efficient `Eval` for the sake of flexibility, or whether the user should copy those upper triangular entries to a full matrix in their code. I do not see a common case where the user should be allowed to call `Eval` when `EvalSymmetric` could be called. The typical usage is in integrators, where the matrix coefficient should be evaluated as efficiently as possible. Do you have other reasons to support the symmetric case in `Eval`?

Reply:
I was mainly thinking of existing integrators.  This new matrix coefficient can only be used with integrators which are aware of the new scenario for handling symmetric coefficients.  It's possible that a user might need to send the same coefficient to two different integrators; a new one they have implemented, and a preexisting one that doesn't know how to handle symmetric coefficients.  This class compels the user to alter existing implementations.

Reply:
Note that the interface changes (e.g. `EvalSymmetric`) were introduced in a recently merged PR, not this PR. This comment is on the `MFEM_VERIFY(!symmetric...` but seems to be about the interface change in general, not just this verify. That is fine, but I just want to clarify that we are discussing more than just this line of code.

@mlstowell For your concern about existing implementations, let's consider how an old integrator would have called `MatrixFunctionCoefficient`. Going back to the July 2, 2020 version of coefficient.hpp in branch `master`, `MatrixFunctionCoefficient` had `Function` and `TDFunction` which could be called by `Eval` to set a `DenseMatrix` (or it could be set to `DenseMatrix mat`, depending on the constructor used). There was no `SymmFunction` or `EvalSymmetric` to set a `Vector` of upper triangular entries. An old integrator would necessarily be using a constant `mat`, or calling `Function` or `TDFunction`. In branch `master` and in this PR, the flag `symmetric` is by default set to `false`. Therefore, `Eval` can still be called as before, and backwards compatibility is not broken.

The only case where `Eval` cannot be called is when the new `MatrixFunctionCoefficient` constructor is called to set `SymmFunction` and `symmetric` to `true`, i.e. the new usage. If this new kind of coefficient is passed to an integrator that calls `Eval`, then it will fail,	and I think this is the issue you are concerned about. In this case, `mat`, `Function`, and `TDFunction` are not set, just `SymmFunction`. At least it will fail with an `MFEM_VERIFY` rather than crashing, so the user has an error message and knows what happened.

This brings us back to the design question of whether we want to (1) force the user to update the integrator code to use `EvalSymmetric` (as in the current state of this PR), or (2) call `SymmFunction` in `Eval` and set the `DenseMatrix` by copying the upper triangular entries to the full matrix. This is a matter of opinion. There is	the choice between (1) which ensures efficient use of `SymmFunction` but is less flexible and user-friendly, and (2) which is flexible but inefficiently copies entries. If you think (2) is better, I can implement the copying.

@tzanio @stefanhenneking @camierjs Do you have a strong preference between (1) and (2)?

Reply:
I do prefer option 2.  Option 1 seems like an "all or nothing" approach.  There may be a middle ground where a user wants to implement their `Function` or `TDFunction` by just computing the upper triangular portion but then use this with an older integrator which expects the full matrix coefficient.  Performing the copy to fill the full matrix may be less than optimal but it's not that expensive and it's easy enough or us to do.

I others have a strong opinion that we should use option 1 I'll accept that and approve.

Reply:
@mlstowell I implemented the copying (option 2 above). There are no tests (unit or otherwise) that use this version (calling Eval for a symmetric function), but the code is short enough that it probably is bug-free.

Reply:
Thanks, @dylan-copeland !